src/Server/GRPCServer.cpp

```diff
@@ -480,6 +481,40 @@ namespace
     };
 
 
+    /// A boolean state protected by mutex able to wait until other thread sets it to a specific value.
+    class BoolState
+    {
+    public:
+        explicit BoolState(bool initial_value) : value(initial_value) {}
+
+        bool get() const
+        {
+            std::lock_guard lock{mutex};
+            return value;
+        }
+
+        void set(bool new_value)
+        {
+            std::lock_guard lock{mutex};
+            if (value == new_value)
+                return;
+            value = new_value;
+            changed.notify_all();
+        }
+
+        void wait(bool wanted_value) const
+        {
+            std::unique_lock lock{mutex};
+            changed.wait(lock, [this, wanted_value]() { return value == wanted_value; });
+        }
+
+    private:
+        bool value;
+        mutable std::mutex mutex;
+        mutable std::condition_variable changed;
```

Comment:
It appears that the previous solution (`std::atomic<bool> value` + null mutex) didn't work very well. Because `changed.wait(lock, predicate)` is the same as

```
while(!predicate())
  changed.wait();
```

and sometimes the following could happen:

```
Thread1 - running wait(): the predicate is checked, and it's false
Thread2 - running set(): value of the predicate is changed, now it's true, but Thread1 don't know yet
Thread2 - running set(): changed.notify_all() is called
Thread1 - running wait(): changed.wait() is called
```

Finally `Thread1` is waiting for `changed` to be notified, but it's never going to happen because it was notified before `changed.wait()` was called.

Reply:
Sure, this is default mistake everyone does using condition variables.